Fix provider type errors in `NetworkController` and its downstream packages

## Explanation

- Resolve type errors in NetworkController and its downstream packages that were marked with `@ts-expect-error TODO:` annotations in https://github.com/MetaMask/core/pull/1653.
- Affected packages: `network-controller`, `selected-network-controller`, `transaction-controller`, `gas-fee-controller`

## Tasks

### Provider type alignment
- `ProxyWithAccessibleTarget<SafeEventEmitterProvider>` (`@metamask/network-controller`) vs. `Provider` (`@metamask/eth-query`)
```ts
 error TS2345: Argument of type 'ProxyWithAccessibleTarget<SafeEventEmitterProvider>' is not assignable to parameter of type 'Provider<JsonRpcParams, Json>'.
  Types of property 'sendAsync' are incompatible.
    Type '(req: JsonRpcRequest<JsonRpcParams>, callback: (error: unknown, providerRes?: any) => void) => void' is not assignable to type '<Params, Result>(payload: SendAsyncPayload<Params>, callback: ProviderSendAsyncCallback<Result>) => void'.
      Types of parameters 'req' and 'payload' are incompatible.
        Type 'SendAsyncPayload<Params>' is not assignable to type 'JsonRpcRequest<JsonRpcParams>'.
```
- `SafeEventEmitterProvider`
- `ProviderProxy`
- `FakeProvider`

## References
- https://github.com/MetaMask/core/issues/1975
- https://github.com/MetaMask/utils/pull/140
- https://github.com/MetaMask/eth-json-rpc-provider/pull/14
- https://github.com/MetaMask/core#workspaces/shared-libraries-621e46b4d7103800171d1b02/issues/zh/244

--- a/packages/assets-controllers/CHANGELOG.md
+++ b/packages/assets-controllers/CHANGELOG.md
@@ -5,6 +5,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+### Changed
+- Bump dependency `@metamask/eth-query` from ^3.0.1 to ^4.0.0 ([#2028](https://github.com/MetaMask/core/pull/2028))
 
 ## [18.0.0]
 ### Changed

--- a/packages/assets-controllers/package.json
+++ b/packages/assets-controllers/package.json
@@ -39,7 +39,7 @@
     "@metamask/base-controller": "^3.2.3",
     "@metamask/contract-metadata": "^2.4.0",
     "@metamask/controller-utils": "^5.0.2",
-    "@metamask/eth-query": "^3.0.1",
+    "@metamask/eth-query": "^4.0.0",
     "@metamask/metamask-eth-abis": "3.0.0",
     "@metamask/network-controller": "^15.2.0",
     "@metamask/polling-controller": "^1.0.1",

--- a/packages/controller-utils/CHANGELOG.md
+++ b/packages/controller-utils/CHANGELOG.md
@@ -5,6 +5,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+### Changed
+- Bump dependency `@metamask/eth-query` from ^3.0.1 to ^4.0.0 ([#2028](https://github.com/MetaMask/core/pull/2028))
 
 ## [5.0.2]
 ### Changed

--- a/packages/controller-utils/package.json
+++ b/packages/controller-utils/package.json
@@ -39,7 +39,7 @@
   },
   "devDependencies": {
     "@metamask/auto-changelog": "^3.4.3",
-    "@metamask/eth-query": "^3.0.1",
+    "@metamask/eth-query": "^4.0.0",
     "@types/jest": "^27.4.1",
     "deepmerge": "^4.2.2",
     "jest": "^27.5.1",

--- a/packages/gas-fee-controller/CHANGELOG.md
+++ b/packages/gas-fee-controller/CHANGELOG.md
@@ -5,6 +5,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+### Changed
+- Bump dependency `@metamask/eth-query` from ^3.0.1 to ^4.0.0 ([#2028](https://github.com/MetaMask/core/pull/2028))
 
 ## [10.0.0]
 ### Added

--- a/packages/gas-fee-controller/package.json
+++ b/packages/gas-fee-controller/package.json
@@ -32,7 +32,7 @@
   "dependencies": {
     "@metamask/base-controller": "^3.2.3",
     "@metamask/controller-utils": "^5.0.2",
-    "@metamask/eth-query": "^3.0.1",
+    "@metamask/eth-query": "^4.0.0",
     "@metamask/network-controller": "^15.2.0",
     "@metamask/polling-controller": "^1.0.1",
     "@metamask/utils": "^8.2.0",

--- a/packages/gas-fee-controller/src/GasFeeController.ts
+++ b/packages/gas-fee-controller/src/GasFeeController.ts
@@ -348,7 +348,6 @@ export class GasFeeController extends PollingController<
     this.legacyAPIEndpoint = legacyAPIEndpoint;
     this.clientId = clientId;
 
-    // @ts-expect-error TODO: Provider type alignment
     this.ethQuery = new EthQuery(this.#getProvider());
 
     if (onNetworkStateChange && getChainId) {
@@ -435,7 +434,6 @@ export class GasFeeController extends PollingController<
       } catch {
         isEIP1559Compatible = false;
       }
-      // @ts-expect-error TODO: Provider type alignment
       ethQuery = new EthQuery(networkClient.provider);
     }
 
@@ -581,7 +579,6 @@ export class GasFeeController extends PollingController<
     const newChainId = networkControllerState.providerConfig.chainId;
 
     if (newChainId !== this.currentChainId) {
-      // @ts-expect-error TODO: Provider type alignment
       this.ethQuery = new EthQuery(this.#getProvider());
       await this.resetPolling();
 

--- a/packages/network-controller/CHANGELOG.md
+++ b/packages/network-controller/CHANGELOG.md
@@ -5,6 +5,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+### Changed
+- Bump dependency `@metamask/eth-query` from ^3.0.1 to ^4.0.0 ([#2028](https://github.com/MetaMask/core/pull/2028))
 
 ## [15.2.0]
 ### Changed

--- a/packages/network-controller/package.json
+++ b/packages/network-controller/package.json
@@ -35,7 +35,7 @@
     "@metamask/eth-json-rpc-infura": "^9.0.0",
     "@metamask/eth-json-rpc-middleware": "^12.0.1",
     "@metamask/eth-json-rpc-provider": "^2.3.0",
-    "@metamask/eth-query": "^3.0.1",
+    "@metamask/eth-query": "^4.0.0",
     "@metamask/json-rpc-engine": "^7.3.0",
     "@metamask/rpc-errors": "^6.1.0",
     "@metamask/swappable-obj-proxy": "^2.1.0",

--- a/packages/network-controller/src/NetworkController.ts
+++ b/packages/network-controller/src/NetworkController.ts
@@ -1013,7 +1013,6 @@ export class NetworkController extends BaseControllerV2<
     }
 
     const networkClient = this.getNetworkClientById(networkClientId);
-    // @ts-expect-error TODO: Provider type alignment
     const ethQuery = new EthQuery(networkClient.provider);
 
     return new Promise((resolve, reject) => {
@@ -1583,7 +1582,6 @@ export class NetworkController extends BaseControllerV2<
       });
     }
 
-    // @ts-expect-error TODO: Provider type alignment
     this.#ethQuery = new EthQuery(this.#providerProxy);
   }
 }

--- a/packages/network-controller/tests/provider-api-tests/helpers.ts
+++ b/packages/network-controller/tests/provider-api-tests/helpers.ts
@@ -462,7 +462,6 @@ export async function withNetworkClient(
 
   const { provider, blockTracker } = clientUnderTest;
 
-  // @ts-expect-error TODO: Provider type alignment
   const ethQuery = new EthQuery(provider);
   const curriedMakeRpcCall = (request: Request) =>
     makeRpcCall(ethQuery, request);

--- a/packages/transaction-controller/CHANGELOG.md
+++ b/packages/transaction-controller/CHANGELOG.md
@@ -5,6 +5,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+### Changed
+- Bump dependency `@metamask/eth-query` from ^3.0.1 to ^4.0.0 ([#2028](https://github.com/MetaMask/core/pull/2028))
 
 ## [16.0.0]
 ### Changed

--- a/packages/transaction-controller/package.json
+++ b/packages/transaction-controller/package.json
@@ -36,7 +36,7 @@
     "@metamask/approval-controller": "^4.1.0",
     "@metamask/base-controller": "^3.2.3",
     "@metamask/controller-utils": "^5.0.2",
-    "@metamask/eth-query": "^3.0.1",
+    "@metamask/eth-query": "^4.0.0",
     "@metamask/gas-fee-controller": "^10.0.0",
     "@metamask/metamask-eth-abis": "^3.0.0",
     "@metamask/network-controller": "^15.2.0",

--- a/packages/transaction-controller/src/TransactionController.ts
+++ b/packages/transaction-controller/src/TransactionController.ts
@@ -435,7 +435,6 @@ export class TransactionController extends BaseController<
     this.provider = provider;
     this.messagingSystem = messenger;
     this.getNetworkState = getNetworkState;
-    // @ts-expect-error TODO: Provider type alignment
     this.ethQuery = new EthQuery(provider);
     this.isSendFlowHistoryDisabled = disableSendFlowHistory ?? false;
     this.isHistoryDisabled = disableHistory ?? false;
@@ -523,7 +522,6 @@ export class TransactionController extends BaseController<
     this.addPendingTransactionTrackerListeners();
 
     onNetworkStateChange(() => {
-      // @ts-expect-error TODO: Provider type alignment
       this.ethQuery = new EthQuery(this.provider);
       this.registry = new MethodRegistry({ provider: this.provider });
       this.onBootCleanup();

